flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java

```diff
@@ -0,0 +1,201 @@
+
+	private static final ObjectMapper mapper = RestMapperUtils.getStrictObjectMapper();
+
+	private final MessageHeaders<R, P, ?> messageHeaders;
+
+	protected AbstractRestHandler(MessageHeaders<R, P, ?> messageHeaders) {
+		this.messageHeaders = messageHeaders;
+	}
+
+	public MessageHeaders<R, P, ?> getMessageHeaders() {
+		return messageHeaders;
+	}
+
+	@Override
+	protected void channelRead0(final ChannelHandlerContext ctx, Routed routed) throws Exception {
+		log.debug("Received request.");
+		final HttpRequest httpRequest = routed.request();
+
+		try {
+			if (!(httpRequest instanceof FullHttpRequest)) {
+				log.error("Implementation error: Received a request that wasn't a FullHttpResponse.");
+				sendErrorResponse(new ErrorResponseBody("Bad request received."), HttpResponseStatus.BAD_REQUEST, ctx, httpRequest);
+				return;
+			}
+
+			ByteBuf msgContent = ((FullHttpRequest) httpRequest).content();
+
+			R request;
+			if (msgContent.capacity() == 0 || messageHeaders.getHttpMethod() == HttpMethodWrapper.GET) {
+				try {
+					request = mapper.readValue("{}", messageHeaders.getRequestClass());
+				} catch (JsonParseException | JsonMappingException je) {
+					log.error("Implementation error: Get request bodies must have a no-argument constructor.", je);
+					sendErrorResponse(new ErrorResponseBody("Internal server error."), HttpResponseStatus.INTERNAL_SERVER_ERROR, ctx, httpRequest);
+					return;
+				}
+			} else {
+				try {
+					ByteBufInputStream in = new ByteBufInputStream(msgContent);
+					request = mapper.readValue(in, messageHeaders.getRequestClass());
+				} catch (JsonParseException | JsonMappingException je) {
+					log.error("Failed to read request.", je);
+					sendErrorResponse(new ErrorResponseBody(String.format("Request did not match expected format %s.", messageHeaders.getRequestClass().getSimpleName())), HttpResponseStatus.BAD_REQUEST, ctx, httpRequest);
+					return;
+				}
+			}
+
+			CompletableFuture<HandlerResponse<P>> response;
+			try {
+				HandlerRequest<R> handlerRequest = new HandlerRequest<>(request, routed.pathParams(), routed.queryParams());
+				response = handleRequest(handlerRequest);
+			} catch (Exception e) {
+				response = FutureUtils.completedExceptionally(e);
+			}
+
+			response.whenComplete((HandlerResponse<P> resp, Throwable error) -> {
+				try {
+					if (error != null) {
+						log.error("Implementation error: Unhandled exception.", error);
+						sendErrorResponse(new ErrorResponseBody("Internal server error."), HttpResponseStatus.INTERNAL_SERVER_ERROR, ctx, httpRequest);
+					} else {
+						if (resp.wasSuccessful()) {
+							sendResponse(messageHeaders, resp.getResponse(), ctx, httpRequest);
+						} else {
+							sendErrorResponse(new ErrorResponseBody(resp.getErrorMessage()), resp.getErrorCode(), ctx, httpRequest);
+						}
+					}
+				} catch (Exception e) {
+					log.error("Critical error while sending a response.", e);
```

Comment:
I think the only reason for an exception here is the failed transformation of the (error)response to a string. If this should happen, then we should still be able to send something to the client. We should let him also know in this case.